roxie/roxiemem/roxiemem.cpp

```diff
@@ -5081,30 +5077,31 @@ class CDataBufferManager : public CInterface, implements IDataBufferManager
         {
             if (curBlock)
             {
-                DataBufferBottom *bottom = (DataBufferBottom *) curBlock;
+                DataBufferBottom *bottom = curBlock;
+                assertex(((memsize_t)bottom & HEAP_PAGE_OFFSET_MASK) == 0);
                 CriticalBlock c(bottom->crit);
                 if (bottom->freeChain)
                 {
                     atomic_inc(&dataBuffersActive);
-                    HeapletBase::link(curBlock);
-                    DataBufferBase *x = bottom->freeChain;
+                    curBlock->Link();
+                    DataBuffer *x = bottom->freeChain;
                     bottom->freeChain = x->next;
                     x->next = NULL;
                     if (memTraceLevel >= 4)
                         DBGLOG("RoxieMemMgr: CDataBufferManager::allocate() reallocated DataBuffer - addr=%p", x);
                     return ::new(x) DataBuffer();
                 }
-                else if ((memsize_t)(nextAddr - curBlock) <= HEAP_ALIGNMENT_SIZE-DATA_ALIGNMENT_SIZE)
+                else if ((((memsize_t)nextAddr) & HEAP_PAGE_OFFSET_MASK) != 0) // Is there any space in the current block
```

Comment:
This feels like it is changing the semantics. Before it was testing whether there was room to add DATA_ALIGNMENT_SIZE, now it is testing if already full ?